src/libraries/Microsoft.PowerFx.Core/Public/FormulaSet.cs

```diff
@@ -0,0 +1,110 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.PowerFx.Core.Utils;
+
+namespace Microsoft.PowerFx.Core.Public
+{
+    /// <summary>
+    /// A set of name/formula paris, which creates and updates a topologically-sorted list
+    /// of name/formula pairs.
+    /// </summary>
+    public class FormulaSet
+    {
+        private readonly IPowerFxEngine _engine;
+
+        private readonly Dictionary<string, FormulaWithParameters> _formulas = new Dictionary<string, FormulaWithParameters>();
+
+        /// <summary>
+        /// A topologically-sorted list of name/formula pairs. This property is regenerated every time
+        /// the formula set is modified.
+        /// </summary>
+        public IEnumerable<KeyValuePair<string, FormulaWithParameters>> SortedFormulas { get; private set; }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="FormulaSet"/> class.
+        /// </summary>
+        public FormulaSet(IPowerFxEngine engine)
+        {
+            _engine = engine;
+            SortFormulas();
+        }
+
+        private void SortFormulas()
+        {
+            var nodes = _formulas.Keys.ToList();
+            var edges = new List<TopologicalSortEdge<string>>();
+
+            foreach (var kvp in _formulas)
+            {
+                var checkResult = _engine.Check(kvp.Value._expression, kvp.Value._schema);
```

Comment:
Do we need to rerun _check on every formula to sort, or can we cache bindings we know haven't changed?

Reply:
We could and we probably should

Reply:
So the call to .Check is now abstracted behind an interface, so if the implementation wants to do caching, it can.